src/ring/middleware/json.clj

```diff
@@ -8,13 +8,20 @@
   (if-let [type (get-in request [:headers "content-type"])]
     (not (empty? (re-find #"^application/(.+\+)?json" type)))))
 
-(defn- read-json [request & [{:keys [keywords? bigdecimals?]}]]
+(defn- ->key-fn [key-fn keywords?]
+  (if (fn? key-fn)
+    (if keywords?
+      #(-> % key-fn keyword)
+      key-fn)
+    keywords?))
+
+(defn- read-json [request & [{:keys [key-fn keywords? bigdecimals?]}]]
   (if (json-request? request)
     (if-let [body (:body request)]
       (let [body-string (slurp body)]
         (binding [parse/*use-bigdecimals?* bigdecimals?]
           (try
-            [true (json/parse-string body-string keywords?)]
+            [true (json/parse-string body-string (->key-fn key-fn keywords?))]
```

Comment:
Instead of `->key-fn`, you could just use `or`, and add a precondition to ensure the options are the right value.